lib/Core/Executor.cpp

```diff
@@ -3552,46 +3574,74 @@ std::string Executor::getAddressInfo(ExecutionState &state,
 }
 
 
-void Executor::terminateState(ExecutionState &state) {
+void Executor::terminateState(ExecutionState &state, StateTerminationType terminationType) {
   if (replayKTest && replayPosition!=replayKTest->numObjects) {
     klee_warning_once(replayKTest,
                       "replay did not consume all objects in test input.");
   }
 
+  // update stats
   interpreterHandler->incPathsExplored();
+  if (processTree)
+    state.ptreeNode->terminationTypeMask |= 1U << static_cast<std::uint32_t>(terminationType);
 
-  std::vector<ExecutionState *>::iterator it =
-      std::find(addedStates.begin(), addedStates.end(), &state);
-  if (it==addedStates.end()) {
-    state.pc = state.prevPC;
+  // exit on termination type
+  if (ExitOnErrorType.isSet(terminationType))
+    haltExecution = true;
 
+  // add state to removedStates for next searcher update
+  auto it = std::find(addedStates.begin(), addedStates.end(), &state);
+  if (it == addedStates.end()) {
+    state.pc = state.prevPC;
     removedStates.push_back(&state);
-  } else {
-    // never reached searcher, just delete immediately
-    std::map< ExecutionState*, std::vector<SeedInfo> >::iterator it3 = 
-      seedMap.find(&state);
-    if (it3 != seedMap.end())
-      seedMap.erase(it3);
-    addedStates.erase(it);
-    processTree->remove(state.ptreeNode);
-    delete &state;
+    return;
   }
+  // ... or just delete immediately if never reached searcher
+  auto it2 = seedMap.find(&state);
+  if (it2 != seedMap.end())
+    seedMap.erase(it2);
+  addedStates.erase(it);
+  if (processTree)
+    processTree->remove(state.ptreeNode);
+  delete &state;
 }
 
-void Executor::terminateStateEarly(ExecutionState &state, 
-                                   const Twine &message) {
-  if (!OnlyOutputStatesCoveringNew || state.coveredNew ||
-      (AlwaysOutputSeeds && seedMap.count(&state)))
-    interpreterHandler->processTestCase(state, (message + "\n").str().c_str(),
-                                        "early");
-  terminateState(state);
+bool shouldWriteTest(const ExecutionState &state, StateTerminationClass terminationClass) {
```

Comment:
You could move this function somewhere else. It does not need the Executor object and you can extern in the options.

Reply:
Where?

Reply:
I was thinking ExecutorUtil, but that doesn't achieve much.  Creating a new file sounds a bit ambitious so maybe ignore this.